libfeed/feed-store-provider.c

```diff
@@ -0,0 +1,1239 @@
+#include "feed-knowledge-app-proxy.h"
+#include "feed-orderable-model.h"
+#include "feed-quote-card-store.h"
+#include "feed-sizes.h"
+#include "feed-store-provider.h"
+#include "feed-text-sanitization.h"
+#include "feed-word-card-store.h"
+#include "feed-word-quote-card-store.h"
+
+typedef GPtrArray * (*ModelsFromResultsAndShardsFunc) (const char * const *shards,
+                                                       GPtrArray          *model_hts,
+                                                       gpointer            user_data);
+typedef GObject * (*ModelFromResultFunc) (GHashTable *model_ht,
+                                          gpointer    user_data);
+
+static GHashTable *
+hashtable_from_variant (GVariant *variant)
+{
+  GVariantIter iter;
+  gchar *key = NULL;
+  gchar *value = NULL;
+  g_autoptr(GHashTable) table = g_hash_table_new_full (g_str_hash,
+                                                       g_str_equal,
+                                                       g_free,
+                                                       g_free);
+
+  g_variant_iter_init (&iter, variant);
+
+  /* Since g_variant_iter_next allocates, we can pass the pointers
+   * directly to g_hash_table_insert without copying their contents */
+  while (g_variant_iter_next (&iter, "{ss}", &key, &value))
+    g_hash_table_insert (table, key, value);
+
+  return g_steal_pointer (&table);
+}
+
+static GPtrArray *
+call_dbus_proxy_and_construct_from_models_and_shards (GDBusProxy                      *proxy,
+                                                      const gchar                     *method_name,
+                                                      const gchar                     *source_name,
+                                                      ModelsFromResultsAndShardsFunc   marshal_func,
+                                                      gpointer                         marshal_data,
+                                                      GCancellable                    *cancellable,
+                                                      GError                         **error)
+{
+  g_autoptr(GPtrArray) model_data_array = NULL;
+  g_autoptr(GVariant) result = g_dbus_proxy_call_sync (proxy,
+                                                       method_name,
+                                                       NULL,
+                                                       G_DBUS_CALL_FLAGS_NONE,
+                                                       -1,
+                                                       cancellable,
+                                                       error);
+  g_autoptr(GVariant) shards_variant = NULL;
+  g_autoptr(GVariant) models_variant = NULL;
+  g_auto(GStrv) shards_strv = NULL;
+  g_autoptr(GPtrArray) model_hts = NULL;
+  GVariantIter iter;
+  GVariant *model_variant = NULL;
+
+  if (result == NULL)
+    return NULL;
+
+  g_variant_get (result, "(^as@aa{ss})", &shards_strv, &models_variant);
```

Comment:
I'm basically just copying what GDbus does here. It might be easier to directly unwrap from `g_variant_get` but I couldn't figure out a good way to do that.